src/vm/gcheaputilities.cpp

```diff
@@ -80,6 +80,105 @@ HMODULE GCHeapUtilities::GetGCModule()
 namespace
 {
 
+// This block of code contains all of the state necessary to handle incoming
+// EtwCallbacks before the GC has been initialized. This is a tricky problem
+// because EtwCallbacks can appear at any time, even when we are just about
+// finished initializing the GC.
+//
+// The below lock is taken by the "main" thread (the thread in EEStartup) and
+// the "ETW" thread, the one calling EtwCallback. EtwCallback may or may not
+// be called on the main thread.
+CRITICAL_SECTION g_eventStashLock;
+
+// These two global fields are 64-bit integers so that loads and stores to them
+// can be atomic.
+LONG64 g_stashedKeywordAndLevel = 0;
+LONG64 g_stashedPrivateKeywordAndLevel = 0;
+
+// In order for loads and stores of the keyword and level state to be atomic,
+// these macros assist in packing and unpacking two 32-bit integer values into
+// a 64-bit value.
+#define STASH_KEY_AND_LEVEL(key, level) ((static_cast<LONG64>(key) << 32) | static_cast<LONG64>(level))
+#define UNSTASH_KEY(combined) (static_cast<GCEventKeyword>(combined >> 32))
+#define UNSTASH_VALUE(combined) (static_cast<GCEventLevel>(combined & 0xFFFFFFFF))
+
+// FinalizeLoad is called by the main thread to complete initialization of the GC.
+// At this point, the GC has provided us with an IGCHeap instance and we are preparing
+// to "publish" it by assigning it to g_pGCHeap.
+//
+// This function can proceed concurrently with StashKeywordAndLevel below.
+void FinalizeLoad(IGCHeap* gcHeap, IGCHandleManager* handleMgr, HMODULE gcModule)
+{
+    // at this point the GC has been initialized and is ready to go.
+    // the main thread is responsible for initializing the lock.
+    UnsafeInitializeCriticalSection(&g_eventStashLock);
+    UnsafeEnterCriticalSection(&g_eventStashLock);
+
+    // the compiler or hardware can't be allowed to reorder this write before
+    // the lock acquision. This is because StashKeywordAndLevel will acquire the
+    // lock if g_pGCHeap is not null and the lock must be initialized by that point.
+    VOLATILE_MEMORY_BARRIER();
+    g_pGCHeap = gcHeap;
+
+    // despite holding the lock, the ETW callback thread may still write to the stashed
+    // values. All writes are atomic.
+    //
+    // the values we read here may not necessarily be the most up-to-date.
+    // Between the time we read these values and we call ControlEvents, an ETW callback
+    // may have been fired and overwrote these values with more up-to-date ones.
+    // However, if this is the case, the ETW callback is now blocked waiting to acquire
+    // g_eventStashLock, so it will have a chance to call ControlEvents itself later
+    // once we release the lock.
+    LONG64 keyAndLevel = InterlockedExchange64(&g_stashedKeywordAndLevel, 0);
```

Comment:
There are too many lock-free operations - it is not easy to see that there are no bugs in it. I would protect it with a simple spin lock (note that spin locks can be initialized as static). Something like:

```
static DangerousNonHostedSpinLock g_eventStashLock;
static BOOL g_gcEventTracingInitialized = FALSE;

void StashKeywordAndLevel(...)
{
    if (!g_gcEventTracingInitialized)
    {
        DangerousNonHostedSpinLockHolder lockHolder(&g_eventStashLock);
        if (!g_gcEventTracingInitialized)
        {
            if (isPublicProvider)
            {
                 g_stashedKeyword = keywords;
                 g_stashedLevel = level;
            }
            else
            {
                 g_stashedPrivateKeyword = keywords;
                 g_stashedPrivateLevel = level;
            }
            return;
        }
    }

    if (isPublicProvider)
    {
        g_pGCHeap->ControlEvents(keywords, level);
    }
    else
    {
        g_pGCHeap->ControlPrivateEvents(keywords, level);
    }
}


void FinalizeLoad(...)
{
    ....

    {
        DangerousNonHostedSpinLockHolder lockHolder(&g_eventStashLock);

        // Ultimately, g_eventStashLock ensures that no two threads call ControlEvents at any
        // point in time.
        g_pGCHeap->ControlEvents(g_stashedKeyword, g_stashedLevel);
        g_pGCHeap->ControlPrivateEvents(g_stashedPrivateKeyword, g_stashedPrivateLevel);

        g_gcEventTracingInitialized = TRUE;
    }
}
```

Reply:
sure - @sergiy-k and I came up with the lock-free strategy but I'm all for simplicity.

Reply:
pushed this to https://github.com/dotnet/coreclr/pull/15873 and rebased this PR against it